src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -280,29 +299,192 @@ def process_input(conf):
+        if not conf.ignore_no_sample_error:
+            raise ValueError(f"Sample data not found for {upas}")
+        return
+
+    with open(metafile, "r", encoding="utf8") as json_file:
+        meta = json.load(json_file)
+
+    upa_dir, sample_file_prefix = Path(metafile).parent, sample_upa.replace("/", "_")
+    sample_file_name = f"{sample_file_prefix}.{loader_common_names.SAMPLE_FILE_EXT}"
+    sample_file = os.path.join(upa_dir, sample_file_name)
+    sample_prepared_name = f"{sample_file_prefix}.{loader_common_names.SAMPLE_PREPARED_EXT}"
+    sample_prepared_file = os.path.join(upa_dir, sample_prepared_name)
+    update_meta = False
+
+    try:
+        if not _check_file_exists(loader_common_names.SAMPLE_FILE_KEY, meta, sample_file):
+            _dump_json_to_file(sample_file, sample_ret)
+            meta[loader_common_names.SAMPLE_FILE_KEY] = sample_file_name
+            update_meta = True
+
+        if not _check_file_exists(loader_common_names.SAMPLE_PREPARED_KEY, meta, sample_prepared_file):
+            node_data = _retrieve_node_data(sample_ret['node_tree'])
+            _dump_json_to_file(sample_prepared_file, node_data)
+            meta[loader_common_names.SAMPLE_PREPARED_KEY] = sample_prepared_name
+            update_meta = True
+    finally:
+        if update_meta:
+            _dump_json_to_file(metafile, meta)
+
+
+def _dump_json_to_file(json_file_path: str, json_data: dict[str, Any]) -> None:
+    with open(json_file_path, "w", encoding="utf8") as json_file:
+        json.dump(json_data, json_file, indent=2)
+
+
+def _find_sample_upa(
+        conf: Conf,
+        upas: list[str]
+) -> (dict[str, Any], str):
+    # find one and only one sample associated upa from input upas and retrieve the sample data
+    # raise error if multiple samples are found
+
+    found_sample, sample_ret, sample_upa = False, None, None
 
-        cfn = os.path.join(conf.job_dir, "workdir/tmp", upa)
-        # upa file is downloaded to cfn
-        conf.asu.get_assembly_as_fasta({"ref": upa.replace("_", "/"), "filename": upa})
+    for upa in upas:
+        try:
+            sample_ret, sample_upa = _retrieve_sample(conf, upa), upa
+            if found_sample:
+                raise ValueError(f"Found multiple samples in input {upas}")
+            found_sample = True
+        except NoDataLinkError:
+            pass
+
+    return sample_ret, sample_upa
+
+
+def _check_file_exists(
+        file_key_name: str,
+        metadata: dict[str, Any],
+        file_path: str
+) -> bool:
+    # check if file exists and if the metadata matches the file name
+
+    return (file_key_name in metadata and
+            metadata[file_key_name] == Path(file_path).name and
+            os.path.isfile(file_path))
+
+
+def _retrieve_sample(conf: Conf, upa: str) -> dict[str, Any] | None:
+    # retrieve sample data from sample service
+
+    # retrieve data links associated with upa
+    links_ret = conf.ss.get_data_links_from_data({"upa": upa})
```

Comment:
sorry, I keep thinking of more complications - for reproducibility, can you add the `effective_time`from this call to the metadata? Since data links can expire or be re-linked we should store the time at which the links were retrieved so we can reproduce the data if needed

Reply:
👍

Reply:
`time.time()` is going to have a different time format than the `effective_time` returned from the server - we should just use the server time

Reply:
👍

Reply:
I mean literally take the value of the `effective_time` key from the return of the `get_data_links_from_data` call and put that in the metadata

Reply:
👍

Reply:
:+1: